### PR TITLE
[ssr] rewrite Ssripats and Ssrview in the tactic monad

### PR DESCRIPTION
**Pull request by gares**
depends on #6676 [merged]
depended upon by #6705

##  ssr: rewrite Ssripats and Ssrview in the tactic monad

`Ssripat`s and `Ssrview` are now written in the Tactic Monad.
`Ssripats` implements the `=>` tactical.
`Ssrview` implements the application of forward views.

The code is, according to my tests, 100% backward compatible.
The code is much more documented than before.
Moreover the `ist` (ltac context) used to interpret views is the correct
one (the one at `ARGUMENT EXTEND` interp time, not the one at `TACTIC
EXTEND` execution time).  Some of the code not touched by this commit
still uses the incorrect `ist`, so its visibility in `TACTIC EXTEND`
can't be removed yet.

The main changes in the code are:
- intro patterns are implemented using a state machine (a goal comes
  with a state). `Ssrcommon.MakeState` provides an easy way for a tactic
  to enrich the goal with with data of interest, such as the set of
  hyps to be cleared.  This cleans up the old implementation that, in
  order to thread the state, that to redefine a bunch of tclSTUFF
- the interpretation of (multiple) forward views uses the state to
  accumulate intermediate results
- the bottom of `Sscommon` collects a bunch of utilities written in the
  tactic monad. Most of them are the rewriting of already existing
  utilities.  When possible the old version was removed.

## Bench
The final bench is at
  https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/343/console
```
┌──────────────────────────┬─────────────────────────┬─────────────────────────────────────┬───────────────────────────────────────┬──────────────────────────┬─────────────────┐
│                          │      user time [s]      │             CPU cycles              │           CPU instructions            │  max resident mem [KB]   │   mem faults    │
│                          │                         │                                     │                                       │                          │                 │
│             package_name │     NEW     OLD PDIFF   │           NEW           OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD  PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│                coq-flocq │   59.28   59.79 -0.85 % │  163360961987  164130189895 -0.47 % │   209817098860   209928779194 -0.05 % │  661724  662028  -0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│              coq-bignums │   75.43   75.60 -0.22 % │  209795361790  210105426772 -0.15 % │   282973447933   283097398279 -0.04 % │  524316  524064  +0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│                 coq-corn │ 1488.82 1491.98 -0.21 % │ 4134652361208 4142647619037 -0.19 % │  6386604329003  6388790265638 -0.03 % │  828000  827896  +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  169.60  169.87 -0.16 % │  471186959143  469528041062 +0.35 % │   660937779861   654968488065 +0.91 % │  649912  587096 +10.70 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│              coq-unimath │ 1197.26 1199.06 -0.15 % │ 3329747275201 3329945987975 -0.01 % │  5739017885925  5734923718832 +0.07 % │  956524 1022956  -6.49 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│                coq-color │  569.62  570.36 -0.13 % │ 1589743770149 1590307788530 -0.04 % │  1979950285536  1980272711262 -0.02 % │ 1437660 1437176  +0.03 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│                  coq-vst │ 2920.91 2922.24 -0.05 % │ 8142944302234 8152735836904 -0.12 % │ 11781160570278 11783642046504 -0.02 % │ 2221056 2220156  +0.04 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   61.05   61.03 +0.03 % │  168790803006  167542227156 +0.75 % │   234364944596   230821564726 +1.54 % │  593120  528676 +12.19 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│               coq-geocoq │ 3045.76 3044.44 +0.04 % │ 8474815873935 8475251963555 -0.01 % │ 13957211859192 13961632218507 -0.03 % │ 1205328 1206040  -0.06 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│      coq-formal-topology │   33.96   33.94 +0.06 % │   94401259874   94469027451 -0.07 % │   123984963326   124055727019 -0.06 % │  481968  481944  +0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  653.72  653.33 +0.06 % │ 1822993805663 1822521509106 +0.03 % │  2946440807755  2945756573347 +0.02 % │ 2972504 2972552  -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│                 coq-hott │  315.86  315.07 +0.25 % │  860207048464  858544147367 +0.19 % │  1374343701278  1374550105658 -0.02 % │  497024  496952  +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│            coq-fiat-core │  102.44  101.99 +0.44 % │  289651316397  289064811919 +0.20 % │   373990504395   374250719307 -0.07 % │  499912  500112  -0.04 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│             coq-compcert │  809.28  805.16 +0.51 % │ 2256060420160 2247041484804 +0.40 % │  3416045599562  3413777934426 +0.07 % │ 1369756 1370160  -0.03 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  443.25  440.75 +0.57 % │ 1232378182451 1225106305951 +0.59 % │  2037560374969  2022361811598 +0.75 % │  812044  717828 +13.13 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│           coq-coquelicot │   76.33   75.84 +0.65 % │  210079002765  207784885497 +1.10 % │   264293519353   258292008807 +2.32 % │  713664  657304  +8.57 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  266.43  264.52 +0.72 % │  739760358681  733145050660 +0.90 % │  1083365378211  1073975712248 +0.87 % │ 1121544  978308 +14.64 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  191.86  190.18 +0.88 % │  531748923791  528800354872 +0.56 % │   764120864069   761578892782 +0.33 % │  848456  708744 +19.71 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1384.46 1370.92 +0.99 % │ 3846233316511 3808422797770 +0.99 % │  6666192488494  6590508255997 +1.15 % │ 1368592 1081716 +26.52 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  168.70  166.62 +1.25 % │  467025749357  462193447032 +1.05 % │   707084547538   699173048256 +1.13 % │  847704  716848 +18.25 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼─────────────────────────────────────┼───────────────────────────────────────┼──────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   41.60   40.44 +2.87 % │  114154621950  111840028077 +2.07 % │   141332091469   137041182630 +3.13 % │  536780  477624 +12.39 % │   0   0  +nan % │
└──────────────────────────┴─────────────────────────┴─────────────────────────────────────┴───────────────────────────────────────┴──────────────────────────┴─────────────────┘
```

### CODE REVIEW

**Comment by gares** [comment]
It seems #6648 will be merged, I'll the rebase and benchmark (I've already started a job, but it includes such fix that imposes a slowdown unrelated to this PR https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/309/)

**Comment by gares** [comment]
@ejgallego is it possible to:
- compile Coq with an f-lambda enabled compiler but with an optimization level equal to the one of standard ocamlc
- compiler ssripats.ml activating all optimizations to get rid of "monadic closures"

I'd like to understand if what we is just the "home grown" monad, or not. I think the code, but for the monadic/CPS style, is pretty similar to the old one.

**Comment by ppedrot** [comment]
@gares Note that using `V82.tactic` is extremely costly in practice, as it entails full normalization of the goal. If anything, I'd start looking there for a source of slowness. The tactic monad only incurs slightly more allocations that shouldn't be responsible for such an increase in time.

**Comment by ejgallego** [comment]
@gares in principle I think that using different flags for files that may be linked in the same unit is not officially supported but I have done it and found no problem in the past, I guess you can try. You could certainly tweak the script to perform the full script with some aggressive flambda options, that should be very easy [just remember to disable the native compiler if you are not using 4.07]

**Comment by gares** [comment]
When I write `foo x >>= fun y -> bar y` don't I allocate more than writing `let y = foo x in bar y`? 

Because a lot of code became of the former kind, eg `tacTYPEOF t >>= fun ty -> ...` and `let tacTYPEOF t = Goal.enter_one fun g -> ... let sigma, ty = Typing.type_of t sigma in tclEVARS sigma <*> tclUNIT ty`.

I think that we have less `V82.tactic` than before.

**Comment by gares** [comment]
In short I write `fun ... -> ... ` way more frequently than before, and if these closure are not eliminated (as flamda does) I have many of them.

**Comment by ppedrot** [comment]
@gares That's true, but it should account for a slow-down of the order of the percent, if not less. These closures are very short-lived and the OCaml GC is pretty efficient in that case.

Judging from your patch, you're calling `V82.tactic` way more often than before, even if there are syntactically less calls... which doesn't seem to be the case anyways, you do seem to have introduced many calls to the compatibility layer. This is definitely known to be source of slowdowns in the range of the 10-20%. So I'd advise writing a variant of that wrapper that does not perform normalization and check whether it changes things. Hopefully the underlying unported code is insensitive to evar normalization...

**Comment by gares** [comment]
I'll try to thinker with that but this is not how you benchmark. We will end up comparing apples with oranges.
When was normalization performed in Coq pre-tactic-monad era, for example?

**Comment by ppedrot** [comment]
@gares Haphazardly in `Tacmach.refine` by sometimes calling `whd_all` or `nf_betaiota`, AFAIU. This was highly unreliable and partial (e.g. it did not do anything to the context I think), and the source of many heisenbugs.

**Comment by gares** [comment]
@ppedrot I removed `nf_evar` from the `V82.tactic` wrappers around elim/case, that is whenever an intro pattern features a `=> […]`. It changes nothing (I even checked again the commits, since I could not believe it).

You said you see more calls to `V82.tactic`. Can you also tell me the most frequent caller? Can you share your profiling data?

**Comment by ppedrot** [comment]
Wait, are we talking about the same piece of code? In the current PR, there are definitely a ton of calls to `V82.tactic` that have been just syntactically introduced...

**Comment by gares** [comment]
I feel like I'm shooting in the dark... I passed `~nf_evars:false` almost everywhere and nothing change.
Are my commits really disabling normalization as I think?

**Comment by ppedrot** [comment]
I must be high on drugs, but I do see a crapload of `V82.tactic` you introduced in the global diff, and you didn't address them in your last patch, only like 3 of those. That's exactly what I meant when I asked about it in the previous thread. Is that some hallucination of mine due to too much tea, or are you the one under the influence of an Alzheimer spell?

**Comment by gares** [comment]
I can't get rid of `V82.tactic`, but I've added a flag `~nf_evars:false` (almost everywhere) that should make the coercion "fast".
Am I missing something?

**Comment by maximedenes** [comment]
> almost everywhere

I see some missing the flag.

File: plugins/ssr/ssrbwd.ml [hunk 1]
```diff
@@ -86,40 +104,55 @@ let mkRAppView ist gl rv gv =
...
+        else [])
+
+let apply_top_tac =
+  Tacticals.tclTHENLIST [
+    introid top_id;
+    apply_rconstr (mkRVar top_id);
+    old_cleartac [SsrHyp(None,top_id)]
+  ]
+
+let inner_ssrapplytac gviews (ggenl, gclr) ist = Proofview.V82.tactic (fun gl ->
```
Comment (maximedenes) [inline plugins/ssr/ssrbwd.ml]: missing the flag

File: plugins/ssr/ssrelim.ml [hunk 2]
```diff
@@ -374,12 +371,14 @@ let ssrelim ?(ind=ref None) ?(is_case=false) ?ist deps what ?elim eqid elim_intr
...
-    Tacticals.tclTHENLIST [refine_with ~with_evars:false elim; cleartac clr] gl in
-  Tacticals.tclTHENLIST [gen_eq_tac; elim_intro_tac ?ist what eqid elim_tac is_rec clr] orig_gl
+    Tacticals.tclTHENLIST [refine_with ~with_evars:false elim; old_cleartac clr] gl in
+  Tacticals.tclTHENLIST [gen_eq_tac; elim_intro_tac what eqid elim_tac is_rec clr] orig_gl
 
 let no_intro ?ist what eqid elim_tac is_rec clr = elim_tac
 
-let elimtac x = ssrelim ~is_case:false [] (`EConstr ([],None,x)) None no_intro
+let elimtac x =
+  Proofview.V82.tactic
```
Comment (maximedenes) [inline plugins/ssr/ssrelim.ml]: missing the flag

**Comment by gares** [comment]
Sure, i ve added the flag step by step, to see if it had impact on soundness and performance.

I will pass the flag to the last 3 calls, buw is it that, so far, nothing changed, and I be added it to many coercion calls

**Comment by gares** [comment]
No change. Looking back at the data, I see ~ 17% *memory* increase.
Isn't that we are just putting more pressure on the GC?

**Comment by maximedenes** [comment]
Did you bench the `state` PR separately, like @ejgallego suggested? That could at least rule out a suspect.

**Comment by ppedrot** [comment]
@gares If you have a specific tactic invocation that is noticeably slower, I can profile it to see what's going wrong.

**Comment by gares** [comment]
it should affect all code, not just ssr. in early benches (that i run on all pkgs) other pks were unaffected IRC

**Comment by maximedenes** [comment]
I see. So maybe like @ppedrot suggested, we should study a concrete example of slowdown.

**Comment by gares** [comment]
Next week I'll try the per-line diff...

**Comment by gares** [comment]
```
Require Import ssreflect.

Lemma xx : nat -> nat.
case  => [| ?].
Show Proof.
```
shows that, without the last commit, the obtained proof term is `... fun _n_ : nat => ?Goal : nat`.
I did not investigate much, but https://github.com/coq/coq/blob/master/tactics/tactics.ml#L779 seems to add a revert cast.  If this is try using `refine` instead of `simple refine` in your tactics does make a copy of the goal (as long as some reduction is in there, I hope, if not always).

**Comment by gares** [comment]
Oh, an BTW, there must be some piece of code that tries hard to print `(?evar : type)` as `?evar`

**Comment by gares** [comment]
Now the bug is that `reduct_in_concl` sees to lose the goal state (e.g. clears are not performed).

**Comment by gares** [comment]
Rebasing over #6648 I've also passed `~typecheck:true`, so a 5% slowdown on solvable is expected

**Comment by gares** [comment]
The good news is that Qed time is back to normal.
The tactics that were slow(er) are still slower, so I imagine I'm still type checking more.
  https://ci.inria.fr/coq/job/benchmark-2-commits/ws/15/html/coq-mathcomp-solvable/mathcomp/solvable/center.v.html

**Comment by gares** [comment]
Go it, we were passing `~typecheck:true` in `unsafe_intro` (while the old code was not doing it, since it calls the standard intro code that does not do it)

**Comment by gares** [comment]
Finally! Time is back to normal. The only thing that increased is memory usage by 16%
@ppedrot is this expected/common porting the code to the TM?

**Comment by gares** [comment]
circle ci reboot

**Comment by gares** [comment]
@maximedenes the bench is now OK! 

Travis fails because I've code like `{ s with x = t }` where s has only one field (it gets more with #6705 that was cut off this PR a posteriori). Is an exception that can be made, or I should fix these ASAP?

**Comment by maximedenes** [comment]
I guess an exception can be made if you can rebase #6705 on top of this PR, so that I can merge both.

**Comment by gares** [comment]
> so that I can merge both.

This PR is, IMHO, quite ready (no review, but the code was already there).  I would not say the same about the other one (little testing, no doc, no review, still some semantic corner cases to think about).

If the exception requires the simultaneous merge, then I'll fix the warnings.

**Comment by maximedenes** [comment]
> If the exception requires the simultaneous merge, then I'll fix the warnings.

I'm afraid so, otherwise it will make Travis fail for all PRs and branches IIUC...

**Comment by herbelin** [comment]
I just had a look at this, considering the questions around #6676. As far as I understand, the main motivation for the added business of #6676 is to have a function `ipat_tac1` of the short type `ssripat -> unit tactic` (*) rather than, say, `ssripat -> to_clear * unit tactic`, or, say, `ssripat -> kont:unit tactic -> unit tactic`, and to use the `Proofview.<*>` (to the price of hard-wiring a new kind of side effect `goal_with_state` in `Proofview`) rather than a custom (and local) `<*>` which would work on type `to_clear * unit tactic` or `kont:unit tactic -> unit tactic` rather than `unit tactic` (compare to the custom and local `tclTHEN_a`-style composition operators of the previous implementation).

I appreciate the effort of the author to think in term of compositionality. The code produced is well-structured and I would be tempted to acknowledge this effort and to recommend the work as it is. I would however recommend for the longer term, that the author try to also think in terms of factorizability of code, which is another important aspect of a sustainable maintenance.

(*) up to the `future_ipats` thing.

**Comment by gares** [comment]
Next time a PR with title "...API..." is merged the week before the freeze I'll just say "do it yourself".

**Comment by maximedenes** [comment]
> I'll just say "do it yourself".

You would say that to whom?

I tried many times to get a schedule (even a rough one) from many developers, @ejgallego included, to anticipate such integration issues, with no luck so far. You're one of the rare exceptions. I would be in favor of having a better collective organization in terms of scheduling, but I don't see how to make that happen.

**Comment by ejgallego** [comment]
> I tried many times to get a schedule (even a rough one) from many developers, @ejgallego included, to anticipate such integration issues, with no luck so far. You're one of the rare exceptions. I would be in favor of having a better collective organization in terms of scheduling, but I don't see how to make that happen.

So the PR that @gares complaints about has been open for almost 3 months, [much older than this one] and now it is my fault? What was I supposed to do? To buy some adsense?

**Comment by gares** [comment]
Sorry guys, but I was a bit pissed of. And I meant "you do the rebase, not me".

I know the PR has been around for long. I've no clue why it took so long so converge. But this is not my point about API PR.

If they are merged soon, they are not going to disrupt other PRs the week before the freeze. If they are not, well, I'd just postpone them. Anyway, sorry for the noise, it's probably nobody's fault.

**Comment by ejgallego** [comment]
@gares I was not complaining about your point of view which I think is very fair. What I don't think is fair is blaming lack of PR coordination on the authors of the PRs.

**Comment by gares** [comment]
I've fixed elpi upstream. I did not check why VST fails now.

**Comment by ejgallego** [comment]
> I've fixed elpi upstream. I did not check why VST fails now.

I think VST is failing from a regression on unrelated universe code, so CI-wise this should be good to go.

**Comment by gares** [comment]
OK, then good night fellas!

**Pull request merged**